### PR TITLE
Adding doc, adding a timer, fixing bugs.

### PR DESCRIPTION
**Pull request by teytaud**
Adding documentation.
Fixing a crash for DCGAN.
Adding a timer.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -64,8 +79,9 @@ And wait for a few days. Your checkpoints will be dumped in output_networks/cele
 For celebaHQ:
 
 ```
-python datasets.py celebaHQ $PATH_TO_CELEBAHQ -o $OUTPUT_DATASET - f
-python train.py PGAN -c config_celebaHQ.json --restart -n celebaHQ
+python datasets.py celebaHQ $PATH_TO_CELEBAHQ -o $OUTPUT_DATASET - f  # Prepare the dataset and build the configuration file.
```
Comment (teytaud) [inline README.md]: not sure of this << - f >> ?

File: README.md [hunk 2]
```diff
@@ -3,7 +3,7 @@
 A GAN toolbox for researchers and developers with:
 - Progressive Growing of GAN(PGAN): https://arxiv.org/pdf/1710.10196.pdf
 - DCGAN: https://arxiv.org/pdf/1511.06434.pdf
-- To come: StyleGAN https://arxiv.org/abs/1812.04948
+- StyleGAN https://arxiv.org/abs/1812.04948
```
Comment (teytaud) [inline README.md]: This is stable, right ?

File: README.md [hunk 3]
```diff
@@ -48,6 +48,21 @@ pip install -r requirements.txt
  - DTD: https://www.robots.ox.ac.uk/~vgg/data/dtd/
  - CIFAR10: http://www.cs.toronto.edu/~kriz/cifar.html
 
+For a quick start with CelebAHQ, you might:
```
Comment (teytaud) [inline README.md]: I just wanted to facilitate people's life by copy-pasting the TLDR.

File: README.md [hunk 4]
```diff
@@ -130,7 +146,8 @@ Where:
 
 1 - MODEL_NAME is the name of the model you want to run. Currently, two models are available:
     - PGAN(progressive growing of gan)
-    - PPGAN(decoupled version of PGAN)
+    - DCGAN
+    - StyleGAN
```
Comment (teytaud) [inline README.md]: I understood correctly ?

File: README.md [hunk 5]
```diff
@@ -209,19 +226,19 @@ You need to use the eval.py script.
 
 You can generate more images from an existing checkpoint using:
 ```
-python eval.py visualization -n $modelName -m $modelType
```
Comment (teytaud) [inline README.md]: there was a confusion between model name, model type and run name --- I propose this version.

File: models/trainer/progressive_gan_trainer.py [hunk 6]
```diff
@@ -22,6 +23,7 @@ def __init__(self,
                  pathdb,
                  miniBatchScheduler=None,
                  datasetProfile=None,
+                 max_time=0,
```
Comment (teytaud) [inline models/trainer/progressive_gan_trainer.py]: @brozi I added a timer, which will help for DFO or AS for GAN optimization.

File: README.md (hunk 4 above)
Comment (teytaud) [reply README.md]: PPGAN and StyleGAN are not there, just PGAN and DCGAN, right ?

File: models/trainer/DCGAN_trainer.py [hunk 7]
```diff
@@ -18,6 +19,7 @@ def getDefaultConfig(self):
 
     def __init__(self,
                  pathdb,
+                 miniBatchScheduler=None,
```
Comment (teytaud) [inline models/trainer/DCGAN_trainer.py]: Adding this argument is necessary because dataset.py sets up miniBatchScheduler and we get a crash at runtime because __init__ does not have such an argument ?

File: README.md (hunk 3 above)
Comment (Molugan) [reply README.md]: Good idea for celebaHQ, it was a pain in the *** to make the first time. 

I would however add a section ## Quick download

File: models/trainer/DCGAN_trainer.py (hunk 7 above)
Comment (Molugan) [reply models/trainer/DCGAN_trainer.py]: It shouldn't crash unless you add a scheduler in the configuration file.

File: README.md [hunk 8]
```diff
@@ -3,7 +3,6 @@
 A GAN toolbox for researchers and developers with:
 - Progressive Growing of GAN(PGAN): https://arxiv.org/pdf/1710.10196.pdf
 - DCGAN: https://arxiv.org/pdf/1511.06434.pdf
-- To come: StyleGAN https://arxiv.org/abs/1812.04948
```
Comment (Molugan) [inline README.md]: :(

**Review by Molugan — DISMISSED** [review]
I'm ok with most changes. Just a little stuff to correct and this can be merged.

File: README.md (hunk 8 above)
Comment (mkondela) [reply README.md]: So no StyleGAN implementation in the end, @Molugan ?😞

File: models/trainer/DCGAN_trainer.py (hunk 7 above)
Comment (teytaud) [reply models/trainer/DCGAN_trainer.py]: Presumably there is a problem in the automatic generation of the configuration file.
@Molugan how should we proceed regarding this issue ? I would go for keeping my fix so that at least it does not crash.

File: README.md (hunk 8 above)
Comment (Molugan) [reply README.md]: There is a PR for styleGAN !

Comment (Molugan) [reply README.md]: styleGAN incoming (https://github.com/facebookresearch/pytorch_GAN_zoo/pull/95) so no need to remove this one

File: README.md (hunk 1 above)
Comment (Molugan) [reply README.md]: -f accelerate the training by saving intermediate images of smaller sizes.

File: models/trainer/DCGAN_trainer.py (hunk 7 above)
Comment (Molugan) [reply models/trainer/DCGAN_trainer.py]: The issue is DCGAN + celebaHQ in the dataset maker. I hadn'y considered the possibility that someone would want to use both.

File: README.md [hunk 9]
```diff
@@ -64,8 +78,9 @@ And wait for a few days. Your checkpoints will be dumped in output_networks/cele
 For celebaHQ:
 
 ```
-python datasets.py celebaHQ $PATH_TO_CELEBAHQ -o $OUTPUT_DATASET - f
-python train.py PGAN -c config_celebaHQ.json --restart -n celebaHQ
+python datasets.py celebaHQ $PATH_TO_CELEBAHQ -o $OUTPUT_DATASET  # Prepare the dataset and build the configuration file.
+python train.py PGAN -c config_celebaHQ.json --restart -n celebaHQ  # Train.
+python eval.py inception -n celebaHQ -m PGAN  # If you want to check the inception score.
```
Comment (Molugan) [inline README.md]: I'd rather not add the inception part here.

File: README.md [hunk 10]
```diff
@@ -250,27 +270,53 @@ You can add optional arguments:
...
 ```
+This feature extractor is then used for computing the similarity.
 
 Then run your model:
 
 ```
-python eval.py inspirational_generation -n $modelName -m $modelType --inputImage $pathTotheInputImage -f $PATH_TO_THE_OUTPUT_FEATURE_EXTRACTOR
+python eval.py inspirational_generation -n $runName -m $modelName --inputImage $pathTotheInputImage -f $PATH_TO_THE_OUTPUT_FEATURE_EXTRACTOR
 ```
+You can compare choose for the optimization one of the optimizers in Nevergrad
```
Comment (Molugan) [inline README.md]: You can compare choose for the optimization one of the optimizers in Nevergrad ->
You can also try out some optimizers from Nevergrad

File: README.md [hunk 11]
```diff
@@ -231,16 +246,21 @@ Will plot a batch of T_SHIRTS in visdom. Please use the option - -showLabels to
...
 
 ### SWD metric
 
 Using the same kind of configuration file as above, just launch:
-
 ```
-python eval.py laplacian_SWD -c $CONFIGURATION_FILE -n $modelName -m $modelType
+python eval.py laplacian_SWD -c $CONFIGURATION_FILE -n $runName -m $modelName
 ```
+for the SWD score, to be maximized, or for the inception score:
```
Comment (Molugan) [inline README.md]: No the SWD should be minimized

File: README.md [hunk 12]
```diff
@@ -250,27 +270,53 @@ You can add optional arguments:
 - -i $ITER: specify the iteration to evaluate(if not set, will take the highest one)
 - --selfNoise: returns the typical noise of the SWD distance for each resolution
 
-### Inspirational generation
+### Inspirational generation (https://arxiv.org/abs/1906.11661)
+
+You might want to generate clothese (or faces, or whatever) using an inspirational image, e.g.:
```
Comment (Molugan) [inline README.md]: typo: clothese -> clothes

I don't find the description very clear.

File: models/trainer/DCGAN_trainer.py (hunk 7 above)
Comment (Molugan) [reply models/trainer/DCGAN_trainer.py]: Then the dataset maker should be changed, not this part.

File: models/trainer/gan_trainer.py [hunk 13]
```diff
@@ -50,6 +51,7 @@ def __init__(self,
             - modelLabel (string): name of the model
             - config (dictionary): configuration dictionnary.
             for all the possible options
+            - max_time (int): max number of seconds for training (0 = infinity).
```
Comment (Molugan) [inline models/trainer/gan_trainer.py]: NIT: working with second is not very practical, hours looks like a better unit to me.

**Review by Molugan — CHANGES_REQUESTED** [review]
The README should be updated. I'm pushing a fix for the dataset maker.
